Update BOUT++ to latest

bendudson opened the pull request:
Fixes compilation issues and upgrades BOUT++ to the `next` branch. This branch includes some important updates, including the `beuler` solver which can be significantly faster at finding steady state solutions than other time integrators.

Also adds a convergence test, automated using Github actions

bendudson committed:
Update to latest BOUT++ next branch

bendudson committed:
Fix compilation errors with latest BOUT++

- InvertPar `Create` -> `create`; return type from raw pointer to `unique_ptr`.
- Ensure that options are read in as Field2D types
- Replace `finite` with `std::isfinite` to squash Apple Clang warnings

bendudson committed:
Update README file

Need `next` branch of BOUT++ to compile now.

bendudson committed:
Adding test for github actions

tests/case-01 performs a simple self-convergence test.
Runs simulations at varying resolution, and tests that the error
converges. Set to fail if order is < 0.5. Result is about 0.8

bendudson committed:
Fixing case-01 example

New version of BOUT++ is more strict with input options:
- Now case sensitive
- Unused options is an error

dschwoerer commented on tests/case-01/runtest:
```diff
@@ -0,0 +1,121 @@
+
+# Python script to run and analyse a self-convergence test
+
+from __future__ import division
+from __future__ import print_function
+
+try:
+  from builtins import str
+except:
+  pass
```
All of this can go if it is for python3

dschwoerer commented on tests/case-01/runtest:
```diff
@@ -0,0 +1,121 @@
+    # Command to run
+    cmd = "../../sd1d "+args
+    # Launch using MPI
+    s, out = launch(cmd, nproc=nproc, pipe=True)
```
use `launch_safe` instead?

dschwoerer commented on tests/case-01/runtest:
```diff
@@ -0,0 +1,121 @@
+from numpy import sqrt, max, abs, mean, array, log, concatenate
+
+# List of NY values to use
+nylist = [20, 40, 80, 160, 320]
```
only the last 3 ones are actually used in the test, not sure what the time overhead is though ...

dschwoerer reviewed (commented):
`black` for the runtest would be nice :-)

bendudson committed:
Improved runtest script

- Black formatted
- Removed two simulations not neede for the test case
- Removed python2 imports
- Use `launch_safe` rather than `launch`

bendudson commented:
Thanks @dschwoerer for all the suggestions! I think I've implemented them all now

bendudson committed:
Fix cases 2 to 5

Input file needs to be changed for BOUT++ v5 / next branch
